aws/retry/middleware.go

```diff
@@ -0,0 +1,164 @@
+package retry
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	middleware2 "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/awslabs/smithy-go/middleware"
+	"github.com/awslabs/smithy-go/transport/http"
+)
+
+// RequestCloner is a function that can take an input request type and clone the request
+// for use in a subsequent retry attempt
+type RequestCloner func(context.Context, interface{}) interface{}
+
+type retryMetadata struct {
+	AttemptNum       int
+	AttemptTime      time.Time
+	MaxAttempts      int
+	AttemptClockSkew time.Duration
+}
+
+type retryMetadataKey struct{}
+
+// RetryMiddleware is a Smithy FinalizeMiddleware that handles retry attempts using the provided
+// Retryer implementation
+type RetryMiddleware struct {
+	retryer       Retryer
+	requestCloner RequestCloner
+}
+
+// NewRetryMiddleware returns a new RetryMiddleware
+func NewRetryMiddleware(retryer Retryer, requestCloner RequestCloner) RetryMiddleware {
+	return RetryMiddleware{retryer: retryer, requestCloner: requestCloner}
+}
+
+func (r RetryMiddleware) ID() string {
+	return "Retry Middleware"
+}
+
+func (r RetryMiddleware) Name() string {
+	return r.ID()
+}
+
+func (r RetryMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (out middleware.FinalizeOutput, metadata middleware.Metadata, err error) {
+	var attemptNum, retryCount int
+	var attemptClockSkew time.Duration
+
+	maxAttempts := r.retryer.MaxAttempts()
+
+	relRetryToken := r.retryer.GetInitialToken()
+
+	origReq := r.requestCloner(ctx, in.Request)
+	origCtx := ctx
+
+	for {
+		attemptNum++
+
+		ctx = setRetryMetadata(origCtx, retryMetadata{
+			AttemptNum:       attemptNum,
+			AttemptTime:      sdk.NowTime(),
+			MaxAttempts:      maxAttempts,
+			AttemptClockSkew: attemptClockSkew,
+		})
+
+		in.Request = r.requestCloner(ctx, origReq)
+
+		out, metadata, reqErr := next.HandleFinalize(ctx, in)
+
+		relRetryToken(reqErr)
+		if reqErr == nil {
+			return out, metadata, nil
+		}
+
+		retryable := r.retryer.IsErrorRetryable(reqErr)
+		if !retryable {
+			return out, metadata, err
+		}
+
+		if maxAttempts > 0 && attemptNum >= maxAttempts {
+			err = &aws.MaxAttemptsError{
+				Attempt: attemptNum,
+				Err:     err,
+			}
+			return out, metadata, err
+		}
+
+		relRetryToken, err = r.retryer.GetRetryToken(ctx, reqErr)
+		if err != nil {
+			return out, metadata, err
+		}
+
+		retryDelay, err := r.retryer.RetryDelay(attemptNum, reqErr)
+		if err != nil {
+			return out, metadata, err
+		}
+
+		if err = sdk.SleepWithContext(ctx, retryDelay); err != nil {
+			err = &aws.RequestCanceledError{Err: err}
+			return out, metadata, err
+		}
+
+		responseMetadata := middleware2.GetResponseMetadata(metadata)
```

Comment:
maybe `smithymiddle` and `awmiddle` for aliases of the two middleware packages?

Reply:
Yeah I forgot to fix the Intellij auto-name